Alerting: Add rule_group label to grafana_alerting_rule_group_rules metric

alexander-akhmetov opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Add `rule_group` label to `grafana_alerting_rule_group_rules` metric (#62361)

---

This PR also addresses the issue where the GroupRules metric (a GaugeVec) keeps its value and is not deleted when an alert rule is removed from the rule registry. 

Previously, when an alert rule with `orgID=1` was active, the metric was:

```
grafana_alerting_rule_group_rules{org="1",state="active"} 1
```

However, after deleting this rule, following calls to `updateRulesMetrics` did not update the gauge value, causing the metric to incorrectly remain at 1.

The fix ensures that when `updateRulesMetrics` is called it also deletes the group rule metrics if needed.

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

alexander-akhmetov committed:
Alerting: Add rule_group label to grafana_alerting_rule_group_rules metric (#62361)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=88289) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=88289) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=88289) it.</sub>

alexander-akhmetov committed:
Alerting: Delete rule group metrics when it is deleted

This commit addresses the issue where the GroupRules metric (a GaugeVec)
keeps its value and is not deleted when an alert rule is removed from the rule registry.
Previously, when an alert rule with orgID=1 was active, the metric was:

  grafana_alerting_rule_group_rules{org="1",state="active"} 1

However, after deleting this rule, subsequent calls to updateRulesMetrics
did not update the gauge value, causing the metric to incorrectly remain at 1.

The fix ensures that when updateRulesMetrics is called it
also deletes the group rule metrics with the corresponding label values if needed.

yuri-tceretian commented on pkg/services/ngalert/schedule/metrics.go:
```diff
@@ -56,20 +72,59 @@ func (sch *schedule) updateRulesMetrics(alertRules []*models.AlertRule) {
+
+// makeRuleGroupLabelValue returns a string that can be used as a label (rule_group) value for alert rule group metrics.
+func makeRuleGroupLabelValue(key models.AlertRuleGroupWithFolderTitleKey) string {
+	return fmt.Sprintf("%s;%s", key.FolderTitle, key.AlertRuleGroupKey.RuleGroup)
```
I am not sure the folder title makes sense here:
- it can contain up to 255 characters
- it can change when a user changes the title.

I think namespace UID would be better here:
- it has a fixed length
- it is mentioned in logs 
- it is used in other places, e.g. RBAC.

alexander-akhmetov replied:
@stevesg [also mentioned some thoughts here](https://github.com/grafana/grafana/issues/62361#issuecomment-2133099465) about using namespace UIDs vs folder titles. One of the main arguments for using folder titles is that Prometheus includes the filename in the same metric. I guess it's also a bit more recognizable from the user's perspective?

@stevesg what do you think?

JohnnyQQQQ commented on pkg/services/ngalert/models/alert_rule.go:
```diff
@@ -428,6 +428,11 @@ type AlertRuleGroupKey struct {
 	RuleGroup    string
 }
 
+type AlertRuleGroupWithFolderTitleKey struct {
```
Nit: I think you might want to change the naming here, so it follows the same pattern as the other keys.
```suggestion
type AlertRuleGroupKeyWithFolderTitle struct {
```

alexander-akhmetov replied:
Thanks, makes sense! Updated: https://github.com/grafana/grafana/pull/88289/commits/77e4283de2e3a6c14a44093c13e4fa2df34bab34

alexander-akhmetov committed:
Code review: rename AlertRuleGroupWithFolderTitleKey to AlertRuleGroupKeyWithFolderTitle

alexander-akhmetov committed:
Rename FolderTitle to FolderFullpath

stevesg reviewed (approved):
Very nice!

diegommm commented:
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

alexander-akhmetov commented:
It's not merged yet because it adds a full folder path to the metric labels, and it would be better to solve https://github.com/grafana/grafana/pull/88376 first.

diegommm commented:
Hi! Greetings from the Release Guild. This is a friendly reminder that the Steady [11.2.x Release Branch](https://github.com/grafana/grafana/tree/v11.2.x) was cut this week, and that it includes the changes committed up to https://github.com/grafana/grafana/commit/b4126d3bce3627e9460e2db0e870c8d984be8e84. This PR was merged after it, so we wanted to let you know that you can add the `backport v11.2.x` label to to backport it and make it part of the next 11.2.x Patch Release, just in case you may have missed the news.

Thank you!
